File: master-thesis-config-2/mtc-2-nodejs-algorithm/src/algorithms/road-plan/logic/population.ts
import Road from './road';
import roadPlanConfig from '../helpers/roadPlanConfig';

class Population {
  roads: Road[];
  maxFitness: number;

  /**
   *      Population constructor
   */
  constructor(roads: Road[]) {
    this.roads = roads;
    this.maxFitness = this.calculateMaxFitness();
  }

  static randomized = (road: Road, size: number) => {
    let tmp: Road[] = [];

    for (let i = 0; i < size; ++i) {
      tmp.push(road.rearrange());
    }

    return new Population(tmp);
  };

  calculateMaxFitness = () => {
    return this.roads.sort((a, b) => b.fitnessRatio - a.fitnessRatio)[0]
      .fitnessRatio;
  };

  selection = () => {
    while (true) {
      var index = Math.floor(
        Math.random() * (roadPlanConfig.populationSize - 1 - 0 + 1) + 0
      );

      if (Math.random() < this.roads[index].fitnessRatio / this.maxFitness) {
        return new Road(this.roads[index].coordinates);
      }
    }
  };

  generateNewPopulation = (size: number) => {
    let roads: Road[] = [];

    for (let i = 0; i < size; ++i) {
      let road = this.selection().crossing(this.selection());

      road.coordinates.forEach((_) => {
        road = road.mutation();
      });

      roads.push(road);
    }

    return new Population(roads);
  };

  getEliteIndividuals = (size: number) => {
    let roads: Road[] = [];
    let tmp = new Population(this.roads);

    for (let i = 0; i < size; ++i) {
      const best = tmp.findBest();

      if (!best) {
        throw new Error('Undefined value for best variable');
      }
      roads.push(best);
      tmp = new Population(
        tmp.roads.filter(
          (p) =>
            roads.find((x) => x.fitnessRatio === p.fitnessRatio) === undefined
        )
      );
    }
    return new Population(roads);
  };

  findBest = () => {
    return this.roads.find((p) => p.fitnessRatio === this.maxFitness);
  };

  evolve = () => {
    let elite = this.getEliteIndividuals(
      roadPlanConfig.numberOfDominantsInNextGeneration
    );
    let newPopulation = this.generateNewPopulation(
      roadPlanConfig.populationSize -
        roadPlanConfig.numberOfDominantsInNextGeneration
    );
    return new Population([...elite.roads, ...newPopulation.roads]);
  };
}

export default Population;

File: master-thesis-config-2/mtc-2-dotnet/Domain/Algorithms/Population.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Domain.Algorithms.Models;

namespace Domain.Algorithms
{
    public class Population
    {
        public List<Road> Roads { get; private set; }
        public double MaxFitness { get; private set; }
        public static Random RandomGenerator { get; set; }

        public Population(List<Road> roads)
        {
            Roads = roads;
            MaxFitness = CalculateMaxFitness();
            RandomGenerator = new Random();
        }

        public static Population Randomized(Road road, int size)
        {
            var tmp = new List<Road>();

            for (var i = 0; i < size; ++i)
                tmp.Add( road.Rearrange() );

            return new Population(tmp);
        }

        private double CalculateMaxFitness() => Roads.Max( t => t.FitnessRatio );

        public Road Selection()
        {
            while (true)
            {
                var index = RandomGenerator.Next(0, Config.populationSize);

                if (RandomGenerator.NextDouble() < Roads[index].FitnessRatio / MaxFitness)
                    return new Road(Roads[index].Coordinates);
            }
        }

        public Population GenerateNewPopulation(int size)
        {
            var roads = new List<Road>();

            for (var i = 0; i < size; ++i)
            {
                var road = Selection().PerformCrossing(Selection());

                foreach (var coord in road.Coordinates)
                    road = road.PerformMutation();

                roads.Add(road);
            }

            return new Population(roads);
        }

        public Population GetEliteIndividuals(int size)
        {
            var roads = new List<Road>();
            var tmp = new Population(Roads);

            for (var i = 0; i < size; ++i)
            {
                roads.Add( tmp.FindBest() );
                tmp = new Population( tmp.Roads.Except(roads).ToList() );
            }

            return new Population(roads);
        }

        public Road FindBest()
        {
            return Roads.FirstOrDefault(t => t.FitnessRatio.CompareTo(MaxFitness) == 0);
        }

        public Population Evolve()
        {
            var elite = GetEliteIndividuals(Config.numberOfDominantsInNextGeneration);
            var newPopulation = GenerateNewPopulation(Config.populationSize - Config.numberOfDominantsInNextGeneration);
            return new Population( elite.Roads.Concat(newPopulation.Roads).ToList() );
        }
    }
}